### PR TITLE
enhancement(loki sink): Define events

### LINKED ISSUE #4981: Define events for the `loki` sink
As part of finishing https://github.com/timberio/vector-product/issues/14, we need to define internal events for the `loki` sink.

## Defining events

* If you have not already, please read [RFC 2064](https://github.com/timberio/vector/blob/master/rfcs/2020-03-17-2064-event-driven-observability.md).
  * ***Be sure to read the [Guide-level Proposal section](https://github.com/timberio/vector/blob/master/rfcs/2020-03-17-2064-event-driven-observability.md#guide-level-proposal) and the implementation details that affect performance.***
* You can also look at the [currently defined events](https://github.com/timberio/vector/tree/master/src/internal_events) for precedence.
* Events should be named properly: `<ComponentName><Noun><Verb>`. For example, the `syslog` source defines the [`SyslogEventReceived` struct](https://github.com/timberio/vector/blob/fb71f5ad76266d9a44164e6804723376353e4ff0/src/internal_events/syslog.rs#L9).

## Requirements

- [ ] Add a `InsertSinkNameHereEventProcessed` event.
  - [x] Derive an `events_processed` metric via the `emit_metrics` method. ([example](https://github.com/timberio/vector/blob/a391434e97a078f6ea92691bba76f9af91ad1866/src/internal_events/add_fields.rs#L8))
- [ ] Add a `InsertSinkNameHereEventDiscarded` event. This event only occurs if the event does not match the filter.
  - [ ] Derive an `events_discarded` metric via the `emit_metrics` method. ([example](https://github.com/timberio/vector/blob/b8403c5eea29ff51df572f7a8c2085d682768744/src/internal_events/sampler.rs#L21))
- [ ] Audit and add events for all possible errors. ([example](https://github.com/timberio/vector/blob/a391434e97a078f6ea92691bba76f9af91ad1866/src/internal_events/lua.rs#L36))
  - [ ] Derive a `processing_errors` metrics via the `emit_metrics` method. ([example](https://github.com/timberio/vector/blob/a391434e97a078f6ea92691bba76f9af91ad1866/src/internal_events/lua.rs#L43))
  - [ ] Emit a log message that ends with `; discarding event` _only_ if the error results in dropping the event. ([example](https://github.com/timberio/vector/blob/a391434e97a078f6ea92691bba76f9af91ad1866/src/internal_events/lua.rs#L39).
- [ ] Audit and add events for all component-specific behavior. ([example](https://github.com/timberio/vector/blob/a391434e97a078f6ea92691bba76f9af91ad1866/src/internal_events/lua.rs#L24))
- [ ] Ensure that logs in the hot path are rate limited via the `rate_limit_secs` parameter. ([example](https://github.com/timberio/vector/blob/fb71f5ad76266d9a44164e6804723376353e4ff0/src/internal_events/tcp.rs#L83)).
- [ ] Ensure that all log messages start with a capital letter and end with a `.`. ([example](https://github.com/timberio/vector/blob/master/src/internal_events/tcp.rs#L83))

### PR DESCRIPTION
**Pull request by ktff**
Closes #4981

Events:
* `events_in_total`
* `events_out_total`   
* `events_discarded_total`   
* `processed_bytes_total`  
* `processing_errors_total` 
* ~new `out_of_order_events_total`~ 
* new `streams_total`		
* ~new `unlabeld_events_total`~ 	
* ~new `large_events_total`~

### CODE REVIEW

File: src/internal_events/loki.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+use super::InternalEvent;
+use metrics::counter;
+
+#[derive(Debug)]
+pub struct LokiEventUnlabeld;
+
+impl InternalEvent for LokiEventUnlabeld {
+    fn emit_metrics(&self) {
+        counter!("unlabeld_events_total", 1);
```
Comment (spencergilbert) [inline src/internal_events/loki.rs:9]:
```suggestion
        counter!("unlabeled_events_total", 1);
```

File: src/internal_events/loki.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
+use super::InternalEvent;
+use metrics::counter;
+
+#[derive(Debug)]
+pub struct LokiEventUnlabeld;
+
+impl InternalEvent for LokiEventUnlabeld {
```
Comment (spencergilbert) [inline src/internal_events/loki.rs:7]:
```suggestion
impl InternalEvent for LokiEventUnlabeled {
```

File: src/internal_events/loki.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
+use super::InternalEvent;
+use metrics::counter;
+
+#[derive(Debug)]
+pub struct LokiEventUnlabeld;
```
Comment (spencergilbert) [inline src/internal_events/loki.rs:5]:
```suggestion
pub struct LokiEventUnlabeled;
```

File: src/sinks/loki.rs [hunk 4]
```diff
@@ -243,6 +244,7 @@ impl HttpSink for LokiSink {
         // `{agent="vector"}` label. This can happen if the only
         // label is a templatable one but the event doesn't match.
         if labels.is_empty() {
+            emit!(LokiEventUnlabeld);
```
Comment (spencergilbert) [inline src/sinks/loki.rs:247]:
```suggestion
            emit!(LokiEventUnlabeled);
```

File: website/cue/reference/components/sinks/loki.cue [hunk 5]
```diff
@@ -201,4 +201,16 @@ components: sinks: loki: {
...
+
+	telemetry: metrics: {
+		events_in_total:           components.sources.internal_metrics.output.metrics.events_in_total
+		events_out_total:          components.sources.internal_metrics.output.metrics.events_out_total
+		events_discarded_total:    components.sources.internal_metrics.output.metrics.events_discarded_total
+		processed_bytes_total:     components.sources.internal_metrics.output.metrics.processed_bytes_total
+		processing_errors_total:   components.sources.internal_metrics.output.metrics.processing_errors_total
+		out_of_order_events_total: components.sources.internal_metrics.output.metrics.out_of_order_events_total
+		streams_total:			   components.sources.internal_metrics.output.metrics.streams_total
+		unlabeld_events_total: 	   components.sources.internal_metrics.output.metrics.unlabeld_events_total
```
Comment (spencergilbert) [inline website/cue/reference/components/sinks/loki.cue:213]:
```suggestion
		unlabele_events_total: 	   components.sources.internal_metrics.output.metrics.unlabeled_events_total
```

File: website/cue/reference/components/sources/internal_metrics.cue [hunk 6]
```diff
@@ -894,6 +912,12 @@ components: sources: internal_metrics: {
 			default_namespace: "vector"
 			tags:              _component_tags
 		}
+		unlabeld_events_total: {
```
Comment (spencergilbert) [inline website/cue/reference/components/sources/internal_metrics.cue:915]:
```suggestion
		unlabeled_events_total: {
```

**Review by spencergilbert — CHANGES_REQUESTED** [review]

File: src/internal_events/batch.rs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+            message = "Event larger than batch max_bytes; dropping event.",
+            batch_max_bytes = %self.max_length,
+            length = %self.length,
+            internal_log_rate_secs = 1
+        );
+    }
+
+    fn emit_metrics(&self) {
+        counter!("events_discarded_total", 1);
+        counter!("large_events_total", 1);
```
Comment (spencergilbert) [inline src/internal_events/batch.rs:22]: I'm not a huge fan of this name, but can't think of anything to replace it with at the moment...

**Review by spencergilbert — COMMENTED** [review]
I'm curious what @bruceg thinks, but I wonder if the metric should just be `events_discarded_total` with a tag for `out-of-order` and `too-large` (ie: a "reason" tag)

File: src/internal_events/loki.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+impl InternalEvent for LokiOutOfOrderEventDropped {
+    fn emit_logs(&self) {
+        warn!(
+            message = "Received out-of-order event; dropping event.",
+            internal_log_rate_secs = 30
+        );
+    }
+
+    fn emit_metrics(&self) {
+        counter!("events_discarded_total", 1);
```
Comment (bruceg) [inline src/internal_events/loki.rs:45]: This should also increment `processing_errors_total`

**Review by bruceg — CHANGES_REQUESTED** [review]
The code looks OK, but there are some naming issues.

File: src/internal_events/batch.rs (hunk 7 above)
Comment (spencergilbert) [reply src/internal_events/batch.rs:22]: 😅 I'll blame the late hour for me, that's a good name.

**Comment by spencergilbert** [comment]
@bruceg @jszwedko any thoughts on the angle that these can be tags on the existing metrics? Rather than new metrics for people to track?

As an operator I'd rather have `total number of events failed` and then drill into _why_. Rather than a number of metric streams that can't be summed (since they're duplicative).

**Comment by ktff** [comment]
Tag sounds fine for `oversize_events_total`, but an out of order event isn't necessarily dropped, if it's configured like it it's timestamp can be overwritten instead.

File: src/internal_events/batch.rs (hunk 7 above)
Comment (ktff) [reply src/internal_events/batch.rs:22]: `oversize_events_total` does sound better.

File: src/internal_events/loki.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+#[derive(Debug)]
+pub struct LokiEventsProcessed {
+    pub byte_size: usize,
+}
+
+impl InternalEvent for LokiEventsProcessed {
+    fn emit_metrics(&self) {
+        counter!("processed_bytes_total", self.byte_size as u64);
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/internal_events/loki.rs:20]: `processed_events_total` are we still tracking that, since we have `events_in_total` and `events_out_total`?

**Comment by ktff** [comment]
But combining what @bruceg said, `out_of_order` could be a part of `processing_errors_total`. As in
```
counter!("processing_errors_total", 1,
            "error_type" => "out_of_order");
```

**Comment by ktff** [comment]
With tags that would be:
```
counter!("events_discarded_total", 1
              "reason" => "oversized");
```
and 
```
counter!("processing_errors_total", 1,
            "error_type" => "out_of_order");
```
and also 
```
counter!("events_discarded_total", 1
              "reason" => "out_of_order");
```
if it does get dropped.

That seems like better approach.

File: src/internal_events/loki.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply src/internal_events/loki.rs:21]: Ah yes, those are centralized.

**Review by bruceg — APPROVED** [review]
LGTM

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**